### ISSUE TITLE
fetch_ticker forces load of all markets

### ISSUE THREAD

**Issue by pursehouse** [ccxt/ccxt#1541]
- OS: all
- Programming Language version: all
- CCXT version:
- Exchange: all
- Method: fetch_ticker

> if I run fetch_ticker ($symbol, $params );

it runs the 

> $this->load_markets();

which means my script that periodically runs will download the entire markets list into memory every time, even though I specifically only need one market. This logic is in every method that is meant to be used on a single market

> fetch_order_book ($symbol, $params = array ()) {
> public function fetch_bid_asks ($symbols = null, $params = array ()) {
> public function fetch_ohlcv ($symbol, $timeframe = '1m', $since = null, $limit = null, $params..

etc etc...

Can this be changed so that it loads the single market when only that is wanted, and leave the global market loading to scenarios this is wanted?
It seems that the logic in **set_markets()** & **load_markets()** to block reloading the entire list when it has been loaded is maintained in the base Exchange.* class unique per language. The variables it uses are sort of confusing, and no code comments isn't helping. But logically the gate should run as knowing the difference between some markets being loaded directly, or the entire market list having already been requested. I see how it is fine if the script using this library is running as a constant process that also for some reason wants every market in memory, but when it's used for a cron that runs on a timer or supplying data on request from a front end UI like a website.. it's going to add a lot to the overhead

**Comment by kroitor** [comment]
```PHP
<?php

$root = dirname (dirname (dirname (__FILE__)));
include $root . '/ccxt.php';
date_default_timezone_set ('UTC');

//-----------------------------------------------------------------------------
// this code snippet will generate PHP code for you
// the generated code will not hit the internet for market loading

$exchange_id = 'gdax'; // set your desired exchange id
$symbol = 'ETH/BTC';   // set your desired market symbol

//-----------------------------------------------------------------------------
// no need to override anything below this line

// instantiate the exchange by id
$exchange = '\\ccxt\\' . $exchange_id;
$exchange = new $exchange ();

// load all markets from the exchange
$markets = $exchange->load_markets ();

// generate the code to preload the chosen symbol
// with exchange without hitting the internet

$market = $exchange->markets[$symbol];
$keys = array ('id', 'symbol', 'base', 'quote');
echo '$id = ' . "'" . $exchange_id . "';\n";
echo '$symbol = ' . "'" . $symbol . "';\n";

// instantiation code
echo '$exchange_id = ' . "'\\\\ccxt\\\\' . " . '$id;' . "\n";
echo '$exchange = new $exchange_id (array ("verbose" => true));' . "\n";

// $this->markets override
echo '$exchange->markets = ' . "array ('" . $market['symbol'] . "' => array (";

$strings = implode (', ', array_map (function ($key) use ($market) {
    return "'" . $key . "' => '" . $market[$key] . "'";
}, $keys));

echo $strings . "));\n";

// $this->markets_by_id override
echo '$exchange->markets_by_id = $exchange->index_by ($exchange->markets, ' . "'id');\n";

// the fetch_ticker call that won't load markets
echo 'var_dump ($exchange->fetch_ticker ($symbol));' . "\n";

?>
```

**Comment by kroitor** [comment]
@pursehouse 

> but doesn't that seem like an extremely complex solution to access what the expected functionality would be for a method that targets a single market?

Let's see... the generated code is:

```PHP
// this line you will need anyway
$exchange_id = 'gdax'; 
$symbol = 'ETH/BTC'; 
$exchange = '\\ccxt\\' . $exchange_id; 
$exchange = new $exchange (array ("verbose" => true)); 

// this is the added lines
$exchange->markets = array ('ETH/BTC' => array ('id' => 'ETH-BTC', 'symbol' => 'ETH/BTC', 'base' => 'ETH', 'quote' => 'BTC')); 
$exchange->markets_by_id = $exchange->index_by ($exchange->markets, 'id');

// this line you will need anyway
var_dump ($exchange->fetch_ticker ($symbol));
```

Right? 

So of 7 lines, you will need 5 of them anyway, and 2 lines were added to override the market loading for you. Does adding 2 lines to your code look extremely complex to you? You gotta be kidding me) Think this way: this isn't complex, this is extremely easy, and if you don't agree, you will have hard time solving any real practical task with it ) 

Those two added lines are the only thing you need, actually. The above code is a little more complex, but I wrote it just to demonstrate how it can be automated for any exchange and market. Most of the time you don't need it and you only need two lines of added code to override them. Does this answer your question?

**Comment by pursehouse** [comment]
right so now those two lines need to be everywhere I access any market data. And for sure people don't realize what is happening internally when they try to get the data for one market. Why not have that functionality inside the system? And without learning the internal working of the system, nobody would ever know to do those commands. And... also it's not getting the data from the source, this requires me to already have all the data of a market. I don't see any methods at all in the system to access a single market from the source.

**Comment by kroitor** [comment]
> Why not have that functionality inside the system? 

If you think of it carefully...

> that functionality

... which functionality ? The functionality of supplying a market of your choice? We don't know which one you want, before you actually tell us. How do you tell us? By supplying a **unified** symbol argument. The unification is done in runtime. We can't know which symbols are offered from the exchange, before we load them. So, how do we know which symbol corresponds to which id? We have to load the desired market once and hardcode it, right? Or save&preload them from a file.

> I don't see any methods at all in the system to access a single market from the source.

If you think of it carefully again, the markets are not accessible from source. Unified markets are created by ccxt for you in runtime. Without having that, you will lose the symbol (pair) unification, you see.

**Comment by pursehouse** [comment]
yes... that's why I said to have it load a single market, not bypass loading entirely.
turn this...
```
public function fetch_ticker ($symbol, $params = array ()) {
        $this->load_markets();
        $market = $this->market ($symbol);
```
into this..
```
public function fetch_ticker ($symbol, $params = array ()) {
        $market = $this->load_market( $symbol );
```
with load market using the same class level storage of the markets as the plural does.

**Comment by kroitor** [comment]
> yes... that's why I said to have it load a single market, not bypass loading entirely.

But 99% of the exchanges don't have the endpoint for loading a single market at all! So you will hit the internet anyway. Isn't this something you were trying to avoid in the first place?

**Comment by kroitor** [comment]
> those two lines need to be everywhere I access any market data

There's no problem to make it a reusable function, right?
Let me know if I still didn't answer your question. Thx!

**Comment by pursehouse** [comment]
no I'm trying to avoid downloading the large list of markets when I only need one.

The one that brought this to my attention because of how long it takes, is CoinMarketCap, which is using the get-all style method to get a single ticker. While that api does have ability to get a single ticker. Binance does as well

**Comment by kroitor** [comment]
Single ticker – yes, single market – no, unfortunately. So you either load them all, or don't load them at all. Does this clarify it now?

**Comment by pursehouse** [comment]
CoinMarketCap

>     public function fetch_markets () {
>         $markets = $this->publicGetTicker (array (
>             'limit' => 0,
>         ));

instead of the single mode

**Comment by pursehouse** [comment]
called from
```
    public function fetch_ticker ($symbol, $params = array ()) {
        $this->load_markets();
```

**Comment by kroitor** [comment]
The last two I didn't quite get, can you clarify what you mean, please?

**Comment by pursehouse** [comment]
it loads every ticker in order to get 1. because "fetch_markets" uses the ticker data

**Comment by kroitor** [comment]
@pursehouse 

> it loads every ticker in order to get 1

Wrong ) It loads all tickers at once in one single request (does not load every ticker separately) – and this is done not to get one ticker, but to get a list of all available markets. And if you don't want a full list, you will have to hardcode your list containing just the market you want. And it won't load it then. Does it make sense now?

**Comment by kroitor** [comment]
@pursehouse ok, let's go the hard way then, if you want ))) 
Let's follow the logic step by step in order for me to be able to show the flaw to you.

1. how do you call the ticker? Can you show a one-liner?

**Comment by pursehouse** [comment]
$currencies = $exchangeApi->fetch_ticker( "origintrail" );

**Comment by kroitor** [comment]
I'm pretty sure you use them differently, because there is no such symbol as "origintrail". What's your exact code, I mean...

**Comment by pursehouse** [comment]
ah right it's. 
`$currencies = $exchangeApi->fetch_ticker( "TRAC/USD" );`
i'd changed things before abandoning using the script and setup instead
`$currencies = json_decode( file_get_contents( "https://api.coinmarketcap.com/v1/ticker/origintrail/" ), true );`

**Comment by pursehouse** [comment]
so perhaps what I'm proposing is having a sub-method that the current methods would pass to using the symbol. so coders could pass symbols to avoid having a redownload of markets
fetch_(whatever)_by_id()

**Comment by kroitor** [comment]
Ok, think where the TRAC/USD coming from. It's a unified symbol. If you don't care about the unification, you can skip loading the entire markets at all, like this:

```
$exchange_id = '\\ccxt\\coinmarketcap';
$exchange = new $exchange_id (array ("verbose" => true));
$ticker = $exchange->public_get_ticker_id (array('convert' => 'USD', 'id' => 'origintrail'));
var_dump($exchange->parse_ticker ($ticker));
```
↑ no markets loaded here, does it answer the question now?

**Comment by pursehouse** [comment]
ok clearly you don't want to keep workarounds instead of extending functionality

**Comment by kroitor** [comment]
@pursehouse but the functionality is already there, isn't it?

> perhaps what I'm proposing is having a sub-method that the current methods would pass to using the symbol. so coders could pass symbols to avoid having a redownload of markets
fetch_(whatever)_by_id()

```PHP
var_dump ($exchange->public_get_ticker_id (array('id' => 'origintrail')));
```

Isn't it exactly it, literally?

**Comment by pursehouse** [comment]
my point is to have all the classes have that ability, not to solve it just for that one specific source

**Comment by kroitor** [comment]
@pursehouse and use exchange-specific non-unified ids with a unified method?

**Comment by pursehouse** [comment]
in your scenario... if I had a database storing the ids, I would also have to store what the method name is to access them. as opposed to universal method like fetch_(whatever)_by_id( $id )

**Comment by kroitor** [comment]
@pursehouse ok, but there are exchanges that don't have any ids and they accept two separate currencies instead, like coinmarketcap does, how would you go around them?

**Comment by pursehouse** [comment]
As in they need the "trac" and "usd" sent to them? Then why would those methods be loading the list of markets if they already have the data

**Comment by kroitor** [comment]
Sorry, can you explain the last argument again, please?

**Comment by pursehouse** [comment]
I think your example is too vague, I don't understand the literal scenario

**Comment by pursehouse** [comment]
ok so if I'm correctly tracking coins in my local compared to the market they belong to, then I am storing their code and the id individually per exchange. Because of the coins ability and existence of having different codes across sources ( BitCoin Cash being BCC/BCH ). So my script knows "TRAC" exists and also knows "origintrail" exists. So ideally the method calls should let me use the known ID instead of making an extra request to the server to get the ID. Instead of currently everything is setup to use the code, and hoping the exchange has that code and checking it after gathering a market list of the entire exchange.
Basically the methods were setup to the symbol which is just a pointer to get to the ID, but they need to know that coin exists on that market anyways, so if it was done with code they would also already know the ID

**Comment by pursehouse** [comment]
so all

> fetch_[anything]( $symbol, .... )
methods should have a matching
> fetch_[anything]_byid( $id, ... )
method which would logically be what the original methods would be using anyways
so the $symbol oriented methods would be a wrapper to the ID oriented methods
example:

>     public function fetch_order_book ($symbol, $params = array ()) {
>         $this->load_markets();
>         $market = $this->market ($symbol);
>         return fetch_order_book_by_id ( $market['id'], $params );
>     }
>     public function fetch_order_book_by_id ( $marketId, $params = array ()) {
>         $orderbook = $this->publicGetDepth (array_merge (array (
>             'symbol' => $marketId,
>             'limit' => 100, // default = maximum = 100
>         ), $params));
>         return $this->parse_order_book($orderbook);
>     }

**Comment by kroitor** [comment]
https://github.com/ccxt/ccxt/issues/1595#issuecomment-361402135

**Comment by andriibezkorovainyi** [comment]
Hey if anyone needed this thing in 2026, here is typescript example with one or multiple single-market uploads with caching
That's usually what you need

```
import type { Exchange, Dictionary, Market } from 'ccxt';
import * as fs from 'node:fs';

export async function loadCcxtMarketsSingleUtil(exchange: Exchange, symbol: string): Promise<void> {
  const path = `./ccxt-markets/${exchange.id}.json`;

  let markets = JSON.parse(fs.readFileSync(path, 'utf-8')) as Dictionary<Market>;

  if (!markets || Object.keys(markets).length === 0) {
    markets = await exchange.loadMarkets();
    fs.writeFileSync(path, JSON.stringify(markets, null));
  }

  const market = markets[symbol];
  if (!market) {
    throw new Error(`Market not found for symbol: ${symbol}`);
  }

  exchange.markets = { ...exchange.markets, symbol: market };
}
```
